MarkBind/markbind
# Implement lazy page building for markbind serve

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [x] Enhancement to an existing feature

Resolves #1007 

**What is the rationale for this request?**

To implement the following two things

#### Lazily building the initial `serve` 
- the loading spinner is shown when the user navigates to pages that have not been built at least once, otherwise the old page is shown while the new page is being built 
- only the landing page is built, taking about ~2.5s

#### Lazily rebuilding affected source files
Editing `text.md` of software engineering topic of 2103 website:
- normal live reloading : 9 pages rebuilt, taking an average of about **7.5s**
- lazy live reloading : 1 page rebuilt, taking around **0.8s**

Gifs are here to keep things lean: https://github.com/MarkBind/markbind/pull/1038#issuecomment-585018267

The main caveat is that heading indexing for unbuilt / unrebuilt pages will not be run, causing fewer than expected search results to show up, until the user navigates to that page.
Hence, the possible search results are limited to the starting page at first, while slowly extending to all pages as the user navigates to them.

In the long run, the savings in rebuild times from rebuilding the page the user is working on only should greatly outweight the "extra time before serving the page".

**What changes did you make? (Give an overview)**
- Repurposed the `-o` option for lazy live reloading, which's behaviour is:
  - only builds the landing page on startup, lazily building other pages as the user navigates to them.
  - lazily builds only the current page if needed, when a file change is detected, and "flags" affected pages for lazily building letter
- Introduced a middleware to detect user requests for a page, allowing processing for that page in the `site` instance:
  - setting the `currentPageViewed`
  - rebuilding the page being navigated to before serving it if : it was never built before, or some file changed before that is a dependency of it.
- Update and organize the cli commands documentation page with some panels
- Introduce a raw html loading spinner page to serve when the page being navigated to needs to be built / rebuilt
- Simple performance logging for live reload timings, including a recommendation message to use lazy loading if the timings crosses certain thresholds 
  - currently, this message repeats itself whenever the threshold is crossed. A seperate PR / issue can look into a package-wide user config of sorts, where this recommendation will be disabled after showing it one time
  - this PR is getting slightly large otherwise, I could implement it here if preferred though


**Provide some example code that this change will affect:**
```js
const lazyReloadMiddleware = function (req, res, next) {
  const isHtmlFile = req.url.endsWith('.html');

  if (isHtmlFile) {
    const isDynamicIncludeHtmlFile = req.url.endsWith('._include_.html');

     if (!isDynamicIncludeHtmlFile) {
       const urlWithoutBaseUrl = req.url.replace(config.baseUrl, '');
       const urlWithoutExtension = fsUtil.removeExtension(urlWithoutBaseUrl);

       // changeCurrentPage initiates the page rebuild if needed
       const didInitiateRebuild = site.changeCurrentPage(urlWithoutExtension);
       if (didInitiateRebuild) {
         // the request is redirected to the loading spinner page in the meantime
         req.url = ensurePosix(path.join(config.baseUrl || '/', LAZY_LOADING_SITE_FILE_NAME));
       }
    }
  }

  next();
};
```

**Is there anything you'd like reviewers to focus on?**
na

**Testing instructions:**
`markbind serve -o <file>` should build only the `<file>` specified ( by default landing page `index.md` if none ) and start the live server.

- Changing any source file that a page is dependent on, and being viewed, should rebuild the page as per normal.
- Changing any source file that a page is dependent on, but not being viewed, should not rebuild the page, but "flag" it to do so later. Also, all page source files except `<file>` should be "flagged" when initially running `markbind serve -o`.
  - Navigating to it later however, should serve the loading page while the page is being rebuilt. The browser should then automatically refresh to show the new page once rebuilt.
- Navigating to pages that are not "flagged" should not show the loading page and should immediately serve the page.

**Proposed commit message: (wrap lines at 72 characters)**
Implement lazy page building for markbind serve

During live reload, changes to a source file rebuilds all files that are
dependent on it.
In addition, rebuilding all dependent source files slows the live reload
process, leading to a less pleasant user experience.

In addition, when running markbind serve, all pages are built and
rendered initially.
This significantly slows down the time to first page load, which can
be substantial when the number of source files is huge.

Let's implement lazy page building in these operations, which is opt-in
using the existing -o <file> or --one-page <file> flags for the markbind
serve command.

Changes to a source file will only rebuild pages that are dependent on
it, and is currently being viewed by the author. Other pages not being
viewed by the author are rebuilt when the author navigates to them.
In addition, only the landing page is built initially when serving.
Subsequent pages are built when the author navigates to them.
